### ISSUE TITLE
Editable install on macOS CI aborts with OMP Error #15 (workaround in place)

### ISSUE THREAD

**Issue by DarshanFofadiya** [DarshanFofadiya/sparselab#18]
## Problem

On GitHub Actions `macos-14` runners, an editable install (`pip install -e .`) of sparselab aborts at import time with:

```
OMP: Error #15: Initializing libomp.dylib, but found libomp.dylib already initialized.
```

Exit code 134 (SIGABRT). The error happens when pytest imports both `torch` and `sparselab` into the same Python process. OpenMP's runtime detects two copies of itself in the process and aborts rather than running with undefined behavior.

Two libomps are present:
1. PyTorch's bundled `torch/lib/libomp.dylib`
2. Our extension's linked Homebrew `libomp.dylib` at `/opt/homebrew/opt/libomp/lib/libomp.dylib`

The `BuildExtWithRepair` class in `setup.py` is supposed to rewrite our `.so`'s install_name to `@rpath/libomp.dylib` and add an rpath pointing at torch's bundled libomp. On my local M3 Pro this works — imports succeed cleanly. On CI's `macos-14` runner the rewrite happens but the loader ends up resolving to Homebrew's libomp anyway.

## Impact — who is affected

- **PyPI wheel users: not affected.** Our `wheels.yml` CI calls `scripts/repair_wheel_macos.sh` post-build, which strips absolute rpaths and uses `@loader_path/../torch/lib` relative references. Shipped wheels on PyPI (v0.2.0, v0.2.1) have this fix applied and work correctly.
- **Editable install users on a fresh machine: maybe affected.** My local M3 Pro happens to work, but `macos-14` CI doesn't. There's an environment variable we don't understand causing the divergence.
- **Editable install on the `macos-14` CI runner in our new `.github/workflows/test.yml`: definitely affected.**

## Current workaround

`test.yml` sets `KMP_DUPLICATE_LIB_OK=TRUE` for the pytest step on macOS only. This tells libomp to continue instead of aborting. Technically allows two independent OpenMP runtimes with separate thread pools, so *theoretical* risk of incorrect parallelism — but for our kernels (which use only `#pragma omp parallel for` on outer row loops), both copies of libomp are the identical LLVM binary, and the workaround is in CI only — never touches user wheels.

## Why this isn't urgent

1. **Shipped wheels aren't affected.** Only editable installs on non-primary-dev machines.
2. **We primarily develop on local macOS where it works.** CI coverage on macOS is nice-to-have; the shipping surface is validated by `wheels.yml`.
3. **Linux x86_64 and Linux aarch64 CI still green.** Those are the two platforms where the AVX2 dW kernel work (issue #2) actually needs validation; macOS is a sibling check.

## Desired fix (not this PR)

Three candidate approaches:

1. **Reproduce locally**: spin up a `macos-14` runner via `act` or a VM; run `otool -L` and `otool -l` on the failing `.so` to see the actual link state; iterate on `BuildExtWithRepair`. ~4 hours.
2. **Debug on CI**: add `otool` output steps to `test.yml`, iterate. Slower per-cycle (~5 min/push) but works. ~8-12 hours wall time across iterations.
3. **Use `-target` or `-Wl,-rpath` differently**: inspect whether GitHub's macos-14 runner sets environment variables that override our rpath. The `ARCHFLAGS=-arch arm64` fix we just applied suggests the runner has universal2-related quirks beyond what we've discovered.

## How to reproduce

1. Have a `macos-14` machine (ideally via GitHub Actions, since my local M3 Pro doesn't repro)
2. `brew install libomp`
3. `pip install -e .`
4. `python -c "import torch, sparselab"` → aborts with OMP error #15

## Related

- CHANGELOG v0.2.0 "editable installs" note — that fix works for local machines, not the GitHub runner
- `scripts/repair_wheel_macos.sh` — the wheel-only repair script that does this correctly for published wheels
- `setup.py`'s `BuildExtWithRepair` class — the editable-install repair that's partially broken on CI

**Comment by DarshanFofadiya** [comment]
## Fixed.

The macOS-arm64 leg of the Test workflow is green for the first time
(441 passed, 93 skipped) and is now a required gate (no
`continue-on-error` masking).

Run on `c7d7ef1`: https://github.com/DarshanFofadiya/sparselab/actions/runs/25523443034

## Root cause

Three things, layered:

1. **`setup.py:configure_openmp()` baked `-Wl,-rpath,/opt/homebrew/opt/libomp/lib` into the `.so`.** That rpath was the only one to survive pip's editable-install pipeline on `macos-14`, so `@rpath/libomp.dylib` resolved to a Homebrew libomp distinct from torch's bundled libomp — two libomps in the process.

2. **macOS does NOT fall back to the flat process namespace when `@rpath/foo` resolves zero rpaths** (my mental model was wrong on the first attempt). The .so's `install_name` must point at a real, on-disk file or a resolvable `@rpath/...` lookup.

3. **`BuildExtWithRepair`'s install_name detector only matched Homebrew prefixes.** Torch's bundled `libomp.dylib` carries `/opt/llvm-openmp/lib/libomp.dylib` as its own install_name; once we pre-installed torch into the runtime env (for `--no-build-isolation`), the linker pulled in torch's libomp and the .so referenced `/opt/llvm-openmp/...` — which the detector ignored, so the rewrite never ran.

## The fix (four commits)

| Commit | What |
|---|---|
| [`96a8f5c`](../commit/96a8f5c) | Drop `-Wl,-rpath,$hb_lib` from `extra_link_args` on macOS. Keep `-L$lib_dir` (link-time only). |
| [`cc76f99`](../commit/cc76f99) | `BuildExtWithRepair` now bakes torch's absolute libomp path into install_name when torch is importable at build time (editable case); falls back to `@rpath/libomp.dylib` otherwise (wheel case, where `repair_wheel_macos.sh` finishes the job). Adds `--no-build-isolation` to the CI install command + canonical `docs/development.md` covering why the flag is required. |
| [`7ed1232`](../commit/7ed1232) | Broaden the install_name detector to match any absolute path ending in `/libomp.dylib`, not just Homebrew prefixes. |
| [`c7d7ef1`](../commit/c7d7ef1) | Cleanup: remove the `KMP_DUPLICATE_LIB_OK=TRUE` workaround, remove `continue-on-error: true` on the macOS leg, fix `test_spmm_rejects_non_cpu_X` to probe MPS allocation instead of trusting `mps.is_available()` (the macos-14 runner advertises MPS but allocation segfaults). |

## Documentation

- New `docs/development.md` explains why `pip install -e '.[dev]' --no-build-isolation` is the canonical editable-install command on every platform, with a troubleshooting section for the "I forgot the flag" symptom.
- `CONTRIBUTING.md` quickstart updated to use the canonical command and links to `docs/development.md`.

## What this does NOT change

- **PyPI wheel users are unaffected.** Wheel builds still go through `scripts/repair_wheel_macos.sh` and that path was already correct.
- **`csrc/`, `sparselab/` Python, autograd, NEON, AVX2, scalar, tests — all untouched** except for the one MPS-probe gate fix.
- **Linux x86_64 + Linux aarch64 paths** are gated on `IS_MACOS` and were never touched.

Closing.
